### DISCUSSION TITLE
Are cif Documents from Structure and read directly from filesystem equal?

### DISCUSSION THREAD

**Discussion by hlvlad** [project-gemmi/gemmi#168] [Q&A] [answered]
Hi,

I would like to ask, whether  data  from original cif file is preserved, when we `read_structure()`. If it is, how can we access it?
```python
block1 = gemmi.read_structure('./3gqc.cif').make_mmcif_document().sole_block()
block1.get_mmcif_category('_pdbx_unobs_or_zero_occ_residues')
# output: empty dict
```
vs
```python
block2 = gemmi.cif.read('./3gqc.cif').sole_block()
block2.get_mmcif_category('_pdbx_unobs_or_zero_occ_residues')
# output: dict with actual data
```

**Edit:**
Actually
```python
len(list(block1))
# output: 48
len(list(block2))
# output: 238
```
Can we create structure from cif document, modify it (say delete some residues) and save new cif with all values of original cif present?

**Comment by wojdyr** [comment, marked as answer]
To preserve Document you can read the structure in two steps:
```py
doc = gemmi.cif.read('./3gqc.cif')
structure = gemmi.make_structure_from_block(doc[0])
```
The same is done internally in `read_structure()`, except that Document is not preserved.
(Update 2023: you can call read_structure() with option `save_doc` to preserve Document.)

Then, after changing the structure, you can modify only selected categories in cif:

```py
groups = gemmi.MmcifOutputGroups(False)  # False -> start with all groups disabled
groups.atoms = True  # enable _atom_site and _atom_site_anisotrop
structure.update_mmcif_block(doc[0], groups)
```

**Reply by lobnaOmar20** [reply]
```
st = gemmi.read_structure('supercell_i00002_w8.cif')
st
<gemmi.Structure # with 0 model(s)>
```
why gemmi.Structure says 0 models ?? 

I tried this instead 
```
doc = gemmi.cif.read('supercell_i00002_w8.cif')
structure = gemmi.make_structure_from_block(doc[0])
```
for neighbor search 
`ns = gemmi.NeighborSearch(doc[0], 1.8)#.populate(include_h=false)`

to get this 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [91], in <cell line: 1>()
----> 1 ns = gemmi.NeighborSearch(doc[0], 1.8)

TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
    1. gemmi.NeighborSearch(model: gemmi.Model, cell: gemmi.UnitCell, max_radius: float)
    2. gemmi.NeighborSearch(st: gemmi.Structure, max_radius: float, model_index: int = 0)
    3. gemmi.NeighborSearch(small_structure: gemmi.SmallStructure, max_radius: float)

Invoked with: <gemmi.cif.Block #>, 1.8

**Reply by wojdyr** [reply]
> ```
> st = gemmi.read_structure('supercell_i00002_w8.cif')
> st
> <gemmi.Structure # with 0 model(s)>
> ```
> 
> why gemmi.Structure says 0 models ??

I suppose that this file is not an mmCIF file with coordinates.

**Reply by lobnaOmar20** [reply]
it is just a cif file with atomic positions for 114 atoms.

**Reply by wojdyr** [reply]
then use function `read_small_structure`
see https://gemmi.readthedocs.io/en/latest/mol.html#small-molecules

**Reply by lobnaOmar20** [reply]
thank you so much for the quick reply.
